From: =?utf-8?b?Ik1hcmNvIFRyZXZpc2FuIChUcmV2acOxbyki?= <mail@3v1n0.net>
Date: Mon, 23 Oct 2023 13:42:06 +0200
Subject: Add support for generic GDM CustomJSON binary protocol

GDM allows passing data through PAM binary conversations, however it
only supports an extension protocol for choice list.

Due to the GDM architecture, creating a new custom protocol is quite
verbose and it implies changing multiple layers, at the same time it is
impossible for modules and GNOME Shell GDM/login extensions to easily
define specific protocols to perform custom actions, and they currently
need to rely on DBus or other ways to communicate without an easy way to
setup a secure communication channel.

At the same time, extending a binary protocol is hard, once one has
been defined.

To make this easier to manage both by modules and by gdm clients, add
a simple JSON-based custom protocol that it's easy to extend and
that moves the protocol definition at higher level than gdm.

This moves the control of custom JSON protocols out from GDM, leaving
to the modules and to gdm clients (the shell) the control over them.

GDM is still in charge of doing some validation check of the JSON
provided by the two parties, but it won't be in charge of any further
protocol operation.
---
 .gitlab-ci.yml                                 |   1 +
 daemon/gdm-session-worker.c                    |  68 +++++++++++
 daemon/gdm-session.c                           | 151 +++++++++++++++++++++++++
 daemon/gdm-session.h                           |   5 +
 daemon/gdm-session.xml                         |  24 ++++
 daemon/meson.build                             |   1 +
 libgdm/gdm-client.c                            |  99 +++++++++++++++-
 libgdm/gdm-client.h                            |   2 +
 meson.build                                    |   8 ++
 pam-extensions/gdm-custom-json-pam-extension.h |  61 ++++++++++
 pam-extensions/gdm-pam-extensions.h            |   2 +
 11 files changed, 419 insertions(+), 3 deletions(-)
 create mode 100644 pam-extensions/gdm-custom-json-pam-extension.h

diff --git a/.gitlab-ci.yml b/.gitlab-ci.yml
index 7cf1b8f..ef28656 100644
--- a/.gitlab-ci.yml
+++ b/.gitlab-ci.yml
@@ -16,6 +16,7 @@ build-fedora:
         gobject-introspection-devel
         gtk3-devel
         iso-codes-devel
+        json-glib-devel
         keyutils-libs-devel
         libXau-devel
         libXdmcp-devel
diff --git a/daemon/gdm-session-worker.c b/daemon/gdm-session-worker.c
index 32f4572..7c2d8c8 100644
--- a/daemon/gdm-session-worker.c
+++ b/daemon/gdm-session-worker.c
@@ -47,6 +47,8 @@
 #include <glib-object.h>
 #include <gio/gio.h>
 
+#include <json-glib/json-glib.h>
+
 #include <X11/Xauth.h>
 
 #include <systemd/sd-daemon.h>
@@ -179,6 +181,7 @@ static char gdm_pam_extension_environment_block[_POSIX_ARG_MAX];
 static const char * const
 gdm_supported_pam_extensions[] = {
         GDM_PAM_EXTENSION_CHOICE_LIST,
+        GDM_PAM_EXTENSION_CUSTOM_JSON,
         NULL
 };
 #endif
@@ -581,6 +584,55 @@ gdm_session_worker_process_choice_list_request (GdmSessionWorker
         return gdm_session_worker_ask_list_of_choices (worker, request->prompt_message, &request->list, &response->key);
 }
 
+static gboolean
+gdm_session_worker_process_custom_json_protocol (GdmSessionWorker            *worker,
+                                                 GdmPamExtensionJSONProtocol *request,
+                                                 GdmPamExtensionJSONProtocol *response)
+{
+        g_autoptr(GError) error = NULL;
+        g_autoptr(JsonParser) parser = NULL;
+        g_autofree char *json_reply = NULL;
+
+        g_debug ("GdmSessionWorker: sending custom JSON protocol request: %s v%d",
+                 request->protocol_name, request->version);
+        g_debug ("GdmSessionWorker: (and waiting for reply)");
+
+        if (!request->json) {
+                g_warning ("GdmSessionWorker: custom JSON request is not valid");
+                return FALSE;
+        }
+
+        parser = json_parser_new_immutable ();
+        if (!json_parser_load_from_data (parser, request->json, -1, &error)) {
+                g_warning ("GdmSessionWorker: custom JSON request is not valid JSON: %s",
+                           error->message);
+                return FALSE;
+        }
+
+        if (!gdm_dbus_worker_manager_call_custom_json_request_sync (worker->manager,
+                                                                    worker->service,
+                                                                    request->protocol_name,
+                                                                    request->version,
+                                                                    request->json,
+                                                                    &response->json,
+                                                                    NULL,
+                                                                    &error)) {
+                g_warning ("GdmSessionWorker: custom JSON request failed: %s",
+                           error->message);
+                return FALSE;
+        }
+
+        if (!response->json) {
+                g_warning ("GdmSessionWorker: custom JSON request returned invalid data");
+                return FALSE;
+        }
+
+        /* No need to validate JSON reply again since that's what we got from
+         * the client and validation happens at daemon level.
+         */
+        return TRUE;
+}
+
 static gboolean
 gdm_session_worker_process_extended_pam_message (GdmSessionWorker          *worker,
                                                  const struct pam_message  *query,
@@ -618,6 +670,22 @@ gdm_session_worker_process_extended_pam_message (GdmSessionWorker          *work
 
                 *response = GDM_PAM_EXTENSION_MESSAGE_TO_PAM_REPLY (list_response);
                 return TRUE;
+        } else if (GDM_PAM_EXTENSION_MESSAGE_MATCH (extended_message, worker->extensions, GDM_PAM_EXTENSION_CUSTOM_JSON)) {
+                GdmPamExtensionJSONProtocol *json_request = (GdmPamExtensionJSONProtocol *) extended_message;
+                g_autofree GdmPamExtensionJSONProtocol *json_response = malloc (GDM_PAM_EXTENSION_CUSTOM_JSON_SIZE);
+
+                g_debug ("GdmSessionWorker: received extended pam message '%s'", GDM_PAM_EXTENSION_CUSTOM_JSON);
+
+                GDM_PAM_EXTENSION_CUSTOM_JSON_RESPONSE_INIT (json_response,
+                                                              json_request->protocol_name,
+                                                              json_request->version);
+
+                if (!gdm_session_worker_process_custom_json_protocol (worker, json_request, json_response)) {
+                        return FALSE;
+                }
+
+                *response = GDM_PAM_EXTENSION_MESSAGE_TO_PAM_REPLY (g_steal_pointer (&json_response));
+                return TRUE;
         } else {
                 g_debug ("GdmSessionWorker: received extended pam message of unknown type %u", (unsigned int) extended_message->type);
                 return FALSE;
diff --git a/daemon/gdm-session.c b/daemon/gdm-session.c
index 06f648e..e5ef594 100644
--- a/daemon/gdm-session.c
+++ b/daemon/gdm-session.c
@@ -43,6 +43,8 @@
 #include <glib-object.h>
 #include <gio/gio.h>
 
+#include <json-glib/json-glib.h>
+
 #include "gdm-session.h"
 #include "gdm-session-glue.h"
 #include "gdm-dbus-util.h"
@@ -774,6 +776,49 @@ gdm_session_handle_choice_list_query (GdmDBusWorkerManager  *worker_manager_inte
         return TRUE;
 }
 
+static gboolean
+gdm_session_handle_custom_json_request (GdmDBusWorkerManager  *worker_manager_interface,
+                                        GDBusMethodInvocation *invocation,
+                                        const char            *service_name,
+                                        const char            *protocol,
+                                        unsigned int           version,
+                                        const char            *request,
+                                        GdmSession            *self)
+{
+        GdmSessionConversation *conversation;
+        GdmDBusUserVerifierCustomJSON *custom_json_interface = NULL;
+
+        g_debug ("GdmSession: custom JSON request for service '%s'", service_name);
+
+        if (self->user_verifier_extensions != NULL) {
+                custom_json_interface =
+                        g_hash_table_lookup (self->user_verifier_extensions,
+                                             gdm_dbus_user_verifier_custom_json_interface_info ()->name);
+        }
+
+        if (custom_json_interface == NULL) {
+                g_dbus_method_invocation_return_error_literal (invocation, G_DBUS_ERROR,
+                                                               G_DBUS_ERROR_NOT_SUPPORTED,
+                                                               "custom JSON interface not supported by client");
+                return TRUE;
+        }
+
+        conversation = find_conversation_by_name (self, service_name);
+        if (conversation != NULL) {
+                set_pending_query (conversation, invocation);
+
+                g_debug ("GdmSession: emitting custom JSON request '%s' v%u",
+                         protocol, version);
+                gdm_dbus_user_verifier_custom_json_emit_request (custom_json_interface,
+                                                                 service_name,
+                                                                 protocol,
+                                                                 version,
+                                                                 request);
+        }
+
+        return TRUE;
+}
+
 static gboolean
 gdm_session_handle_info_query (GdmDBusWorkerManager  *worker_manager_interface,
                                GDBusMethodInvocation *invocation,
@@ -1268,6 +1313,11 @@ export_worker_manager_interface (GdmSession      *self,
                                  G_CALLBACK (gdm_session_handle_choice_list_query),
                                  self,
                                  0);
+        g_signal_connect_object (worker_manager_interface,
+                                 "handle-custom-json-request",
+                                 G_CALLBACK (gdm_session_handle_custom_json_request),
+                                 self,
+                                 0);
 
         g_dbus_interface_skeleton_export (G_DBUS_INTERFACE_SKELETON (worker_manager_interface),
                                           connection,
@@ -1303,6 +1353,9 @@ unexport_worker_manager_interface (GdmSession           *self,
         g_signal_handlers_disconnect_by_func (worker_manager_interface,
                                               G_CALLBACK (gdm_session_handle_choice_list_query),
                                               self);
+        g_signal_handlers_disconnect_by_func (worker_manager_interface,
+                                              G_CALLBACK (gdm_session_handle_custom_json_request),
+                                              self);
 }
 
 static gboolean
@@ -1392,6 +1445,81 @@ export_user_verifier_choice_list_interface (GdmSession      *self,
                              interface);
 }
 
+static gboolean
+gdm_session_handle_client_custom_json_reply (GdmDBusUserVerifierCustomJSON *custom_json_interface,
+                                             GDBusMethodInvocation         *invocation,
+                                             const char                    *service_name,
+                                             const char                    *json,
+                                             GdmSession                    *self)
+{
+        g_autoptr(GError) error = NULL;
+        g_autoptr(JsonParser) parser = NULL;
+
+        g_debug ("GdmSession: user replied with custom JSON");
+
+        parser = json_parser_new_immutable ();
+        if (!json_parser_load_from_data (parser, json, -1, &error)) {
+                g_autofree char *message = NULL;
+
+                message = g_strdup_printf ("JSON reply is not valid: %s", error->message);
+                g_warning ("GdmSession: %s", message);
+
+                g_dbus_method_invocation_return_error_literal (invocation,
+                                                               G_DBUS_ERROR,
+                                                               G_DBUS_ERROR_NOT_SUPPORTED,
+                                                               message);
+                gdm_session_report_error (self, service_name,
+                                          G_DBUS_ERROR_NOT_SUPPORTED,
+                                          message);
+                return TRUE;
+        }
+
+        gdm_dbus_user_verifier_custom_json_complete_reply (custom_json_interface, invocation);
+        gdm_session_answer_query (self, service_name, json);
+        return TRUE;
+}
+
+static gboolean
+gdm_session_handle_client_custom_json_report_error (GdmDBusUserVerifierCustomJSON *custom_json_interface,
+                                                    GDBusMethodInvocation         *invocation,
+                                                    const char                    *service_name,
+                                                    const char                    *message,
+                                                    GdmSession                    *self)
+{
+        g_debug ("GdmSession: user reported custom JSON error: %s", message);
+
+        gdm_dbus_user_verifier_custom_json_complete_report_error (custom_json_interface, invocation);
+        gdm_session_report_error (self, service_name, G_DBUS_ERROR_ACCESS_DENIED, message);
+        return TRUE;
+}
+
+static void
+export_user_verifier_custom_json_interface (GdmSession      *self,
+                                             GDBusConnection *connection)
+{
+        GdmDBusUserVerifierCustomJSON *interface;
+
+        interface = GDM_DBUS_USER_VERIFIER_CUSTOM_JSON (gdm_dbus_user_verifier_custom_json_skeleton_new ());
+
+        g_signal_connect (interface,
+                          "handle-reply",
+                          G_CALLBACK (gdm_session_handle_client_custom_json_reply),
+                          self);
+        g_signal_connect (interface,
+                          "handle-report-error",
+                          G_CALLBACK (gdm_session_handle_client_custom_json_report_error),
+                          self);
+
+        g_dbus_interface_skeleton_export (G_DBUS_INTERFACE_SKELETON (interface),
+                                          connection,
+                                          GDM_SESSION_DBUS_OBJECT_PATH,
+                                          NULL);
+
+        g_hash_table_insert (self->user_verifier_extensions,
+                             gdm_dbus_user_verifier_custom_json_interface_info ()->name,
+                             interface);
+}
+
 static gboolean
 gdm_session_handle_client_enable_extensions (GdmDBusUserVerifier    *user_verifier_interface,
                                              GDBusMethodInvocation  *invocation,
@@ -1411,6 +1539,10 @@ gdm_session_handle_client_enable_extensions (GdmDBusUserVerifier    *user_verifi
                             gdm_dbus_user_verifier_choice_list_interface_info ()->name) == 0)
                         export_user_verifier_choice_list_interface (self, connection);
 
+                if (g_str_equal (extensions[i],
+                                 gdm_dbus_user_verifier_custom_json_interface_info ()->name))
+                        export_user_verifier_custom_json_interface (self, connection);
+
         }
 
         gdm_dbus_user_verifier_complete_enable_extensions (user_verifier_interface, invocation);
@@ -3127,6 +3259,25 @@ gdm_session_answer_query (GdmSession *self,
         }
 }
 
+void
+gdm_session_report_error (GdmSession *self,
+                          const char *service_name,
+                          GDBusError  code,
+                          const char *message)
+{
+        GdmSessionConversation *conversation;
+
+        g_return_if_fail (GDM_IS_SESSION (self));
+        g_return_if_fail (service_name != NULL);
+
+        conversation = find_conversation_by_name (self, service_name);
+        if (conversation == NULL)
+                return;
+
+        g_dbus_method_invocation_return_error_literal (g_steal_pointer (&conversation->pending_invocation),
+                                                       G_DBUS_ERROR, code, message);
+}
+
 void
 gdm_session_cancel  (GdmSession *self)
 {
diff --git a/daemon/gdm-session.h b/daemon/gdm-session.h
index dc1eeef..0f05a6a 100644
--- a/daemon/gdm-session.h
+++ b/daemon/gdm-session.h
@@ -22,6 +22,7 @@
 #define __GDM_SESSION_H
 
 #include <glib-object.h>
+#include <gio/gio.h>
 #include <sys/types.h>
 
 G_BEGIN_DECLS
@@ -115,6 +116,10 @@ void              gdm_session_close                       (GdmSession *session);
 void              gdm_session_answer_query                (GdmSession *session,
                                                            const char *service_name,
                                                            const char *text);
+void              gdm_session_report_error                (GdmSession *session,
+                                                           const char *service_name,
+                                                           GDBusError  code,
+                                                           const char *message);
 void              gdm_session_select_program              (GdmSession *session,
                                                            const char *command_line);
 void              gdm_session_select_session              (GdmSession *session,
diff --git a/daemon/gdm-session.xml b/daemon/gdm-session.xml
index 137be5e..8b5115c 100644
--- a/daemon/gdm-session.xml
+++ b/daemon/gdm-session.xml
@@ -28,6 +28,14 @@
       <arg name="query" direction="in" type="a{ss}"/>
       <arg name="answer" direction="out" type="s"/>
     </method>
+    <method name="CustomJSONRequest">
+      <annotation name="org.gtk.GDBus.C.Name" value="CustomJsonRequest" />
+      <arg name="service_name" direction="in" type="s"/>
+      <arg name="protocol" direction="in" type="s"/>
+      <arg name="version" direction="in" type="u"/>
+      <arg name="request" direction="in" type="s"/>
+      <arg name="response" direction="out" type="s"/>
+    </method>
   </interface>
   <interface name="org.gnome.DisplayManager.UserVerifier">
     <method name="EnableExtensions">
@@ -95,6 +103,22 @@
       <arg name="list" type="a{ss}"/>
     </signal>
   </interface>
+  <interface name="org.gnome.DisplayManager.UserVerifier.CustomJSON">
+    <method name="Reply">
+      <arg name="service_name" direction="in" type="s"/>
+      <arg name="reply" direction="in" type="s"/>
+    </method>
+    <method name="ReportError">
+      <arg name="service_name" direction="in" type="s"/>
+      <arg name="error" direction="in" type="s"/>
+    </method>
+    <signal name="Request">
+      <arg name="service_name" type="s"/>
+      <arg name="protocol" type="s"/>
+      <arg name="version" type="u"/>
+      <arg name="request" type="s"/>
+    </signal>
+  </interface>
   <interface name="org.gnome.DisplayManager.Greeter">
     <method name="SelectSession">
       <arg name="session" direction="in" type="s"/>
diff --git a/daemon/meson.build b/daemon/meson.build
index 41f30ab..c3a4c73 100644
--- a/daemon/meson.build
+++ b/daemon/meson.build
@@ -60,6 +60,7 @@ gdm_daemon_deps = [
   gobject_dep,
   gio_dep,
   gio_unix_dep,
+  json_glib,
   libpam_dep,
   x_deps,
   xcb_dep,
diff --git a/libgdm/gdm-client.c b/libgdm/gdm-client.c
index 66c13a8..451ada6 100644
--- a/libgdm/gdm-client.c
+++ b/libgdm/gdm-client.c
@@ -228,6 +228,35 @@ on_user_verifier_choice_list_proxy_created (GObject            *source,
         maybe_complete_user_verifier_proxy_operation (client, data);
 }
 
+static void
+on_user_verifier_custom_json_proxy_created (GObject            *source,
+                                            GAsyncResult       *result,
+                                            UserVerifierData   *data)
+{
+        GHashTable *user_verifier_extensions;
+        g_autoptr(GdmClient) client = NULL;
+        GdmUserVerifierCustomJSON *custom_json;
+        g_autoptr(GError) error = NULL;
+
+        client = GDM_CLIENT (g_async_result_get_source_object (G_ASYNC_RESULT (data->task)));
+
+        user_verifier_extensions = g_object_get_qdata (G_OBJECT (data->user_verifier),
+                                                       gdm_client_user_verifier_extensions_quark ());
+        custom_json = gdm_user_verifier_custom_json_proxy_new_finish (result, &error);
+
+        if (custom_json == NULL) {
+                g_warning ("Couldn't create UserVerifier CustomJSON proxy: %s", error->message);
+                g_hash_table_remove (user_verifier_extensions,
+                                     gdm_user_verifier_custom_json_interface_info ()->name);
+        } else {
+                g_hash_table_replace (user_verifier_extensions,
+                                      gdm_user_verifier_custom_json_interface_info ()->name,
+                                      custom_json);
+        }
+
+        maybe_complete_user_verifier_proxy_operation (client, data);
+}
+
 static void
 on_user_verifier_extensions_enabled (GdmUserVerifier    *user_verifier,
                                      GAsyncResult       *result,
@@ -271,6 +300,17 @@ on_user_verifier_extensions_enabled (GdmUserVerifier    *user_verifier,
                                                                  (GAsyncReadyCallback)
                                                                  on_user_verifier_choice_list_proxy_created,
                                                                  data);
+                } else if (g_str_equal (client->enabled_extensions[i],
+                                        gdm_user_verifier_custom_json_interface_info ()->name)) {
+                        g_hash_table_insert (user_verifier_extensions, client->enabled_extensions[i], NULL);
+                        gdm_user_verifier_custom_json_proxy_new (connection,
+                                                                 G_DBUS_PROXY_FLAGS_NONE,
+                                                                 NULL,
+                                                                 SESSION_DBUS_PATH,
+                                                                 cancellable,
+                                                                 (GAsyncReadyCallback)
+                                                                 on_user_verifier_custom_json_proxy_created,
+                                                                 data);
                 } else {
                         g_debug ("User verifier extension %s is unsupported", client->enabled_extensions[i]);
                         g_hash_table_remove (user_verifier_extensions,
@@ -730,6 +770,20 @@ gdm_client_get_user_verifier_sync (GdmClient     *client,
                                                                                                                               NULL);
                                                         if (choice_list_interface != NULL)
                                                                     g_hash_table_insert (user_verifier_extensions, client->enabled_extensions[i], choice_list_interface);
+                                            } else if (g_str_equal (client->enabled_extensions[i],
+                                                       gdm_user_verifier_custom_json_interface_info ()->name)) {
+                                                        GdmUserVerifierCustomJSON *custom_json_interface;
+                                                        custom_json_interface = gdm_user_verifier_custom_json_proxy_new_sync (connection,
+                                                                                                                                G_DBUS_PROXY_FLAGS_NONE,
+                                                                                                                                NULL,
+                                                                                                                                SESSION_DBUS_PATH,
+                                                                                                                                cancellable,
+                                                                                                                                NULL);
+                                                        if (custom_json_interface != NULL) {
+                                                                g_hash_table_insert (user_verifier_extensions,
+                                                                                     client->enabled_extensions[i],
+                                                                                     custom_json_interface);
+                                                        }
                                             }
                                 }
                         }
@@ -836,6 +890,24 @@ gdm_client_get_user_verifier_finish (GdmClient       *client,
         return user_verifier;
 }
 
+static GHashTable *
+get_user_verifier_extensions (GdmClient *client)
+{
+        GHashTable *user_verifier_extensions = NULL;
+
+        if (client->user_verifier_for_reauth != NULL) {
+                user_verifier_extensions = g_object_get_qdata (G_OBJECT (client->user_verifier_for_reauth),
+                                                               gdm_client_user_verifier_extensions_quark ());
+        }
+
+        if (user_verifier_extensions == NULL && client->user_verifier != NULL) {
+                user_verifier_extensions = g_object_get_qdata (G_OBJECT (client->user_verifier),
+                                                               gdm_client_user_verifier_extensions_quark ());
+        }
+
+        return user_verifier_extensions;
+}
+
 /**
  * gdm_client_get_user_verifier_choice_list:
  * @client: a #GdmClient
@@ -849,7 +921,7 @@ gdm_client_get_user_verifier_finish (GdmClient       *client,
 GdmUserVerifierChoiceList *
 gdm_client_get_user_verifier_choice_list (GdmClient *client)
 {
-        GHashTable *user_verifier_extensions = NULL;
+        GHashTable *user_verifier_extensions = get_user_verifier_extensions (client);
 
         if (client->user_verifier_for_reauth != NULL) {
                 user_verifier_extensions = g_object_get_qdata (G_OBJECT (client->user_verifier_for_reauth),
@@ -861,11 +933,31 @@ gdm_client_get_user_verifier_choice_list (GdmClient *client)
                                                                gdm_client_user_verifier_extensions_quark ());
         }
 
+        return g_hash_table_lookup (user_verifier_extensions,
+                                    gdm_user_verifier_choice_list_interface_info ()->name);
+}
+
+/**
+ * gdm_client_get_user_verifier_custom_json:
+ * @client: a #GdmClient
+ *
+ * Gets a #GdmUserVerifierCustomJSON object that can be used to
+ * verify a user's local account.
+ *
+ * Returns: (transfer none): #GdmUserVerifierCustomJSON or %NULL if user
+ * verifier isn't yet fetched, or daemon doesn't support the custom JSON
+ * protocol
+ */
+GdmUserVerifierCustomJSON *
+gdm_client_get_user_verifier_custom_json (GdmClient *client)
+{
+        GHashTable *user_verifier_extensions = get_user_verifier_extensions (client);
+
         if (user_verifier_extensions == NULL)
                 return NULL;
 
         return g_hash_table_lookup (user_verifier_extensions,
-                                    gdm_user_verifier_choice_list_interface_info ()->name);
+                                    gdm_user_verifier_custom_json_interface_info ()->name);
 }
 
 static void
@@ -1433,7 +1525,8 @@ gdm_client_new (void)
  * @extensions: (array zero-terminated=1) (element-type utf8): a list of extensions
  *
  * Enables GDM's pam extensions.  Currently, only
- * org.gnome.DisplayManager.UserVerifier.ChoiceList is supported.
+ * `org.gnome.DisplayManager.UserVerifier.ChoiceList` and
+ * `org.gnome.DisplayManager.UserVerifier.CustomJSON` are supported.
  */
 void
 gdm_client_set_enabled_extensions (GdmClient          *client,
diff --git a/libgdm/gdm-client.h b/libgdm/gdm-client.h
index f73b555..95f9fcb 100644
--- a/libgdm/gdm-client.h
+++ b/libgdm/gdm-client.h
@@ -70,6 +70,8 @@ GdmUserVerifier   *gdm_client_get_user_verifier_sync    (GdmClient *client,
 
 GdmUserVerifierChoiceList *gdm_client_get_user_verifier_choice_list  (GdmClient *client);
 
+GdmUserVerifierCustomJSON *gdm_client_get_user_verifier_custom_json (GdmClient *client);
+
 void               gdm_client_get_greeter               (GdmClient     *client,
                                                          GCancellable         *cancellable,
                                                          GAsyncReadyCallback   callback,
diff --git a/meson.build b/meson.build
index bf0e585..358d39c 100644
--- a/meson.build
+++ b/meson.build
@@ -43,6 +43,12 @@ config_h_dir = include_directories('.')
 udev_dep = dependency('udev')
 gudev_dep = dependency('gudev-1.0', version: '>= 232')
 
+# PAM
+libpam_dep = cc.find_library('pam')
+pam_extensions_supported = cc.has_header_symbol(
+  'security/pam_appl.h', 'PAM_BINARY_PROMPT',
+  dependencies: libpam_dep)
+
 glib_min_version = '2.56.0'
 
 glib_dep = dependency('glib-2.0', version: '>=' + glib_min_version)
@@ -50,6 +56,8 @@ gobject_dep = dependency('gobject-2.0', version: '>=' + glib_min_version)
 gio_dep = dependency('gio-2.0', version: '>=' + glib_min_version)
 gio_unix_dep = dependency('gio-unix-2.0', version: '>=' + glib_min_version)
 gtk_dep = dependency('gtk+-3.0', version: '>= 2.91.1')
+json_glib = dependency('json-glib-1.0', version: '>= 1.2.0',
+  required: pam_extensions_supported)
 libcanberra_gtk_dep = dependency('libcanberra-gtk3', version: '>= 0.4')
 accountsservice_dep = dependency('accountsservice', version: '>= 0.6.35')
 xcb_dep = dependency('xcb')
diff --git a/pam-extensions/gdm-custom-json-pam-extension.h b/pam-extensions/gdm-custom-json-pam-extension.h
new file mode 100644
index 0000000..a50ae12
--- /dev/null
+++ b/pam-extensions/gdm-custom-json-pam-extension.h
@@ -0,0 +1,61 @@
+/* -*- Mode: C; tab-width: 8; indent-tabs-mode: nil; c-basic-offset: 8 -*-
+ *
+ * Copyright (C) 2023 Canonical Ltd.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ * Author: Marco Trevisan (Treviño) <marco.trevisan@canonical.com>
+ *
+ */
+
+#pragma once
+
+#include "gdm-pam-extensions-common.h"
+
+typedef struct {
+        GdmPamExtensionMessage header;
+
+        const char protocol_name[64];
+        unsigned int version;
+        char *json;
+} GdmPamExtensionJSONProtocol;
+
+#define GDM_PAM_EXTENSION_CUSTOM_JSON "org.gnome.DisplayManager.UserVerifier.CustomJSON"
+#define GDM_PAM_EXTENSION_CUSTOM_JSON_SIZE sizeof (GdmPamExtensionJSONProtocol)
+
+#define GDM_PAM_EXTENSION_CUSTOM_JSON_REQUEST_INIT(request, proto_name, proto_version, json_str) \
+{ \
+        size_t proto_len = strnlen ((proto_name), sizeof ((request)->protocol_name) - 1); \
+        GDM_PAM_EXTENSION_LOOK_UP_TYPE (GDM_PAM_EXTENSION_CUSTOM_JSON, &((request)->header.type)); \
+        (request)->header.length = htobe32 (GDM_PAM_EXTENSION_CUSTOM_JSON_SIZE); \
+        memcpy ((char *)(request)->protocol_name, (proto_name), proto_len); \
+        ((char *)((request)->protocol_name))[proto_len] = '\0'; \
+        (request)->version = (proto_version); \
+        (request)->json = (char *) (json_str); \
+}
+
+#define GDM_PAM_EXTENSION_CUSTOM_JSON_RESPONSE_INIT(response, proto_name, proto_version) \
+{ \
+        size_t proto_len = strnlen ((proto_name), sizeof ((response)->protocol_name) - 1); \
+        GDM_PAM_EXTENSION_LOOK_UP_TYPE (GDM_PAM_EXTENSION_CUSTOM_JSON, &((response)->header.type)); \
+        (response)->header.length = htobe32 (GDM_PAM_EXTENSION_CUSTOM_JSON_SIZE); \
+        memcpy ((char *)(response)->protocol_name, (proto_name), proto_len); \
+        ((char *)((response)->protocol_name))[proto_len] = '\0'; \
+        (response)->version = (proto_version); \
+        (response)->json = NULL; \
+}
+
+#define GDM_PAM_EXTENSION_REPLY_TO_CUSTOM_JSON_RESPONSE(reply) \
+        ((GdmPamExtensionJSONProtocol *) (void *) reply->resp)
diff --git a/pam-extensions/gdm-pam-extensions.h b/pam-extensions/gdm-pam-extensions.h
index cdf34f3..9df8e37 100644
--- a/pam-extensions/gdm-pam-extensions.h
+++ b/pam-extensions/gdm-pam-extensions.h
@@ -1,6 +1,7 @@
 /* -*- Mode: C; tab-width: 8; indent-tabs-mode: nil; c-basic-offset: 8 -*-
  *
  * Copyright (C) 2017 Red Hat, Inc.
+ * Copyright (C) 2023 Canonical Ltd.
  *
  * This program is free software; you can redistribute it and/or modify
  * it under the terms of the GNU General Public License as published by
@@ -21,5 +22,6 @@
 #define GDM_PAM_EXTENSIONS_H
 
 #include "gdm-choice-list-pam-extension.h"
+#include "gdm-custom-json-pam-extension.h"
 
 #endif
